src/python/gem5/simulate/multi_sim.py

```diff
@@ -0,0 +1,113 @@
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import itertools
+import os
+import signal
+from functools import partial
+from time import sleep
+from typing import (
+    Any,
+    Callable,
+    List,
+    Tuple,
+)
+
+from gem5.resources.resource import AbstractResource
+from gem5.utils.multiprocessing import Process
+
+from .simulator import Simulator
+
+
+class MultiSim:
+    """
+    This class is used to run multiple simulations in parallel. It takes a list of
+    callables that return a Simulator object and runs them in parallel. The number of
+    simulations that can be run in parallel is limited to the value of the num_cpus parameter
+    """
+
+    def __init__(self, sim_callables: List[Callable[[], Simulator]]):
+        """
+        :param sim_callables: A list of callables that return a Simulator object
+        """
+        self.sim_callables = sim_callables
+
+    def run_all(self, num_cpus: int):
+        """
+        Run all the simulations in parallel
+        :param num_cpus: The number of simulations to run in parallel at a time
+        """
+        processes = []
+
+        for sim_callable in self.sim_callables:
+            if len(processes) > num_cpus:
+                for process in processes:
+                    if not process.is_alive():
+                        processes.remove(process)
+                    os.kill(process.pid, signal.SIGTERM)
+                sleep(1)
+            process_name = (
+                f"{sim_callable.func.__name__}_{sim_callable.args[0].get_id()}"
+            )
+            process = Process(
+                target=run_simulator,
+                args=(sim_callable,),
+                name=process_name,
+            )
+            print(f"Starting process {process_name}")
+
+            process.start()
+            processes.append(process)
+
+        while processes:
+            for process in processes:
+                if not process.is_alive():
+                    processes.remove(process)
+            sleep(1)
+
+        print("All simulations have finished")
+
+
+def get_cross_product_sim_callables(
+    configs: List[Any], resources: List[AbstractResource]
+):
+    sim_callables = []
+
+    for config, resource in itertools.product(configs, resources):
+        sim_callables.append(partial(config, resource))
+
+    return sim_callables
```

Comment:
I think it's best to leave it up to the user to define their partial functions. I suspect if you try to handle every use case like this we'll end up with far too many functions to properly maintain.

So i'd just remove this helper function.

Reply:
I *think* I follow what you're saying Bobby, but I'm not totally certain. Are you suggesting deleting `get_cross_product_sim_callable`?

It could be useful to have some "common" examples here, but I agree that we can't be everything to everyone. I'm also not sure how to use this particular function since it has no documentation.

Harshil, either remove the function or add documentation.